pkg/actions/mustgather.go

```diff
@@ -0,0 +1,364 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+
+package actions
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/eclipse/codewind-installer/pkg/appconstants"
+	"github.com/eclipse/codewind-installer/pkg/docker"
+	"github.com/eclipse/codewind-installer/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+var codewindHome = filepath.Join(homeDir, ".codewind")
+var nowTime = time.Now().Format("20060102150405")
+var mustGatherDirName = filepath.Join(codewindHome, "mustgather", nowTime)
+
+var isLoud = true
+
+func logMG(input ...string) {
+	if isLoud {
+		fmt.Println(input[0])
+	}
+}
+
+//MustGatherCommand to gather logs and project files to aid diagnosis of Codewind errors
+func MustGatherCommand(c *cli.Context) {
+	if c.Bool("quiet") {
+		isLoud = false
+	}
+	dirErr := os.MkdirAll(filepath.Join(mustGatherDirName, "projects"), 0755)
+	if dirErr != nil {
+		errors.CheckErr(dirErr, 205, "")
+	}
+	logMG("Mustgather files will be collected at " + mustGatherDirName)
+
+	// Collect Codewind container inspection & logs
+	for _, cwContainerName := range docker.ContainerNames {
+		logMG("Collecting information from container " + cwContainerName)
+		containerID := getContainerID(cwContainerName)
+		writeContainerInspectToFile(containerID, cwContainerName)
+		writeContainerLogToFile(containerID, cwContainerName)
+	}
+
+	// Collect Codewind PFE workspace
+	logMG("Collecting Codewind workspace")
+	pfeContainerID := getContainerID(docker.PfeContainerName)
+	copyCodewindWorkspace(pfeContainerID)
+
+	if c.Bool("projects") {
+		// Collect project container inspection & logs
+		dockerClient, dockerErr := docker.NewDockerClient()
+		if dockerErr != nil {
+			HandleDockerError(dockerErr)
+			os.Exit(1)
+		}
+		allContainers, cListErr := docker.GetContainerListWithOptions(dockerClient, types.ContainerListOptions{All: true})
+		if cListErr != nil {
+			HandleDockerError(cListErr)
+			os.Exit(1)
+		}
+		for _, cwContainer := range docker.GetContainersToRemove(allContainers) {
```

Comment:
I'd be very tempted to rename `docker.GetContainersToRemove` now to something like `docker.GetCodewindProjectContainers` since we're using it not just for the list of containers to stop.

Reply:
Agreed